From ce2b3fa8d71f832e99a1c9d1eb38f6e1cc39b1a2 Mon Sep 17 00:00:00 2001
From: Ulrich Obergfell <uobergfe@redhat.com>
Date: Mon, 28 Apr 2014 13:58:16 -0400
Subject: [drm] nouveau/bios: fix a bit shift error introduced by recent commit

Message-id: <1398693496-19424-2-git-send-email-uobergfe@redhat.com>
Patchwork-id: 79198
O-Subject: [RHEL7 BZ1089936 PATCH 1/1] drm/nouveau/bios: fix a bit shift error introduced by RHEL7 commit 24a4279
Bugzilla: 1089936
RH-Acked-by: Ben Skeggs <bskeggs@redhat.com>
RH-Acked-by: Rob Clark <rclark@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

This is a backport of the below upstream commit.

commit 9a11843987b29b7b7eff650bff7d5731ba5117d5
Author: Sergei Antonov <saproj@gmail.com>
Date:   Tue Apr 15 23:18:37 2014 +0200

    drm/nouveau/bios: fix a bit shift error introduced by 457e77b

    Commit 457e77b26428ab4a24998eecfb99f27fa4195397 added two checks applied to a
    value received from nv_rd32(bios, 0x619f04). But after this new piece of code
    is executed, the addr local variable does not hold the same value it used to
    hold before the commit. Here is what is was assigned in the original code:
     (u64)(nv_rd32(bios, 0x619f04) & 0xffffff00) << 8
    in the committed code it ends up with this value:
     (u64)(nv_rd32(bios, 0x619f04) >> 8) << 8
    These expressions are obviously not equivalent.

    My Nvidia video card does not show anything on the display when I boot a
    kernel containing this commit.

    The patch fixes the code so that the new checks are still done, but the
    side effect of an incorrect addr value is gone.

    Cc: Ben Skeggs <bskeggs@redhat.com>
    Cc: Dave Airlie <airlied@redhat.com>
    Cc: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Sergei Antonov <saproj@gmail.com>
    Signed-off-by: Dave Airlie <airlied@redhat.com>
Signed-off-by: Ulrich Obergfell <uobergfe@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/nouveau/core/subdev/bios/base.c b/drivers/gpu/drm/nouveau/core/subdev/bios/base.c
index aac99e4..cc3ae2e 100644
--- a/drivers/gpu/drm/nouveau/core/subdev/bios/base.c
+++ b/drivers/gpu/drm/nouveau/core/subdev/bios/base.c
@@ -105,7 +105,7 @@ nouveau_bios_shadow_pramin(struct nouveau_bios *bios)
    return;
   }
 
-  addr = (u64)(addr >> 8) << 8;
+  addr = (addr & 0xffffff00) << 8;
   if (!addr) {
    addr  = (u64)nv_rd32(bios, 0x001700) << 16;
    addr += 0xf0000;
-- 
1.7.1